### PR TITLE
[Hold][Feature] File quotas

### PR DESCRIPTION
**Pull request by chrisseto**
See commit messages for details.

~~Waiting on text for emails and the final say on the UI.~~

### CODE REVIEW

**Comment by chrisseto** [comment]
@lyndsysimon or @sloria ready for first pass.

File: tests/webtest_tests.py [hunk 1]
```diff
@@ -85,7 +85,7 @@ def test_sees_projects_in_her_dashboard(self):
         res = res.click('My Dashboard', index=0, auth=self.user.auth)
         assert_in('Projects', res)  # Projects heading
 
-    def test_does_not_see_osffiles_in_user_addon_settings(self):
+    def test_not_does_sees_osffiles_in_user_addon_settings(self):
```
Comment (lyndsysimon) [inline tests/webtest_tests.py]: Not sure why this was renamed?

File: website/addons/osfstorage/model.py [hunk 2]
```diff
@@ -24,6 +32,124 @@
...
+        )
+
+        if save:
+            self.storage_usage = usage
+            self.save()
+
+        return usage
+
+    def __repr__(self):
+        return '<{}({!r}, Using {:.2f}%)>'.format(self.__class__.__name__, self.owner, self.storage_usage / self.storage_limit * 100)
```
Comment (lyndsysimon) [inline website/addons/osfstorage/model.py]: :heart:

File: website/addons/osfstorage/tests/utils.py [hunk 3]
```diff
@@ -37,8 +37,11 @@ def setUp(self):
         super(StorageTestCase, self).setUp()
 
         self.project = ProjectFactory()
+        self.node = self.project  # Less typing
```
Comment (lyndsysimon) [inline website/addons/osfstorage/tests/utils.py]: :heart:

**Comment by lyndsysimon** [comment]
First pass complete.

Gold star, just for you!

:star:

**Comment by chrisseto** [comment]
@lyndsysimon or @sloria Updated with develop and the proposal.
Ready for final code review and merge pending @bnosek's approval.

**Comment by bnosek** [comment]
Do I need to review the emails we drafted to make approval?  Or, something else as well?

**Comment by chrisseto** [comment]
@bnosek, just to confirm that the proposal is finished and there are no unresolved comments/questions.
Feel free to review the language if you'd like, another set of eyes if never a bad thing.

**Comment by bnosek** [comment]
text of emails looks okay to me.

**Comment by lyndsysimon** [comment]
Consider my code review complete - note that I have not completed a functional review, nor have I compared functionality with the proposal.

I'm happy to do so if necessary, but in my opinion this is ready to be merged and deployed to staging for testing.

**Comment by sloria** [comment]
ping @GaryKriebel @lbanner @erinmayhood for functional review.

**Comment by lbanner** [comment]
OK

**Comment by GaryKriebel** [comment]
@sloria, for the functional review, I assume I need to run the file_quotas_migration.py script and perhaps lower the limit in website/addons/osfstorage/__init__.py for ease of testing?

**Comment by sloria** [comment]
@GaryKriebel correct.

```
python -m scripts.osfstorage.file_quotas_migration
```

**Comment by GaryKriebel** [comment]
@chrisseto ... I apologize in advance for the length of this comment, but I think the best way is via screenshots.  There seem to be a few issues with the file quotas.

First, I have not seen any emails in the consoles.  Not sure if that part is working or not.

Second, there seem to be some math issues.  I changed my local defaults.py as such:
screen shot 2015-08-07 at 2 28 53 pm
 
I wiped my DB and created my user and have only 1 project (output of the migration script):
screen shot 2015-08-07 at 2 23 46 pm

I uploaded 5 files:
screen shot 2015-08-07 at 2 24 42 pm

Ran the script (math is off):
screen shot 2015-08-07 at 2 25 33 pm

I deleted the files:
screen shot 2015-08-07 at 2 25 51 pm

Ran the script .... should have 0 bytes:
screen shot 2015-08-07 at 2 26 18 pm

Uploaded again:
screen shot 2015-08-07 at 2 26 37 pm

Ran the script:
screen shot 2015-08-07 at 2 27 56 pm

Uploaded 5 more:
screen shot 2015-08-07 at 2 28 11 pm

Ran the script:
screen shot 2015-08-07 at 2 29 13 pm

Made a folder, uploaded 10 more into that folder .... and it let me
screen shot 2015-08-07 at 2 29 45 pm

forgot to capture the script.... upload 1 more file ... get the notice that I'm over my limit... good...

Delete the folder with the 10 files it it:
screen shot 2015-08-07 at 2 31 34 pm

Delete the 10 files in the root storage:
screen shot 2015-08-07 at 2 32 10 pm

Run the script... I owe OSF Storage 2 gigs?
screen shot 2015-08-07 at 2 32 32 pm

**Comment by chrisseto** [comment]
@GaryKriebel, my apologies I forgot to mention that a change to Waterbutler is required for this feature. I'll Make sure that using the proper branch of WB fixes the issues you've been seeing and then update the original post.

**Comment by GaryKriebel** [comment]
I think I pulled one from your repo, but I'll double check and try again.

**Comment by GaryKriebel** [comment]
@chrisseto ... almost there?

The node is being updated when adding and deleting files (I have not checked moving in and out of OSF storage), but the user is not.  I added 5 100mb files then deleted the files.

screen shot 2015-08-11 at 10 17 35 am
screen shot 2015-08-11 at 10 17 05 am
screen shot 2015-08-11 at 10 16 48 am
screen shot 2015-08-11 at 10 16 10 am
screen shot 2015-08-11 at 10 14 47 am

Also, a few questions... does the migration script get run every night to update storage and send emails (I did see the email text when passing the threshold) or does another script handle that?  If I get close to the limit, will I as a user get an email every time that script is run until I am under the limit?  The proposal doesn't seem to address how many times a user close to the limit will receive the email.

**Comment by GaryKriebel** [comment]
(I updated the comment above to reflect the correct order of the screenshots... first pass had them backwards)

**Comment by chrisseto** [comment]
I ran in to this issue yesterday, somehow an incorrect field is making its
way into mongo and odm is overriding a property. It may be a side effect of
the earlier migration script or something more odd could be happening.

You'll have to remove the key "storage_usage" from the collection
"osfstoragenodesettings".

If my hunch is correct the aforementioned key will be set to 0 on all
entries.

**Comment by GaryKriebel** [comment]
Where is "osfstoragenodesettings"?

Also, the proposal states that the same files uploaded to two or more projects will only be counted once.

I started blank:
screen shot 2015-08-11 at 10 38 12 am

Uploaded the same 12 files to Project 1 and Project 2.
screen shot 2015-08-11 at 10 39 25 am
screen shot 2015-08-11 at 10 39 39 am

My usage is the addition of the two:
screen shot 2015-08-11 at 10 40 09 am

**Comment by GaryKriebel** [comment]
in model.py

    root_node = fields.ForeignField('OsfStorageFileNode')
    file_tree = fields.ForeignField('OsfStorageFileTree')

    # The current storage being used by this user (in osfstorage) in bytes
    storage_usage = fields.IntegerField(default=0)

    # Temporary field to mark that a record has been migrated by the
    # migrate_from_oldels scripts
    _migrated_from_old_models = fields.BooleanField(default=False)


so comment out the following?
    storage_usage = fields.IntegerField(default=0)

**Comment by chrisseto** [comment]
My apologies that comment should have read "osfstoragenodesettings", its a
bit early over here.
It's not in the code but just an extraneous value in the database.

The proposal was a bit vague but if a new version is not created, uploading
a file with the same contents and same name, that will not be counted
against you.

**Comment by GaryKriebel** [comment]
Well... as seen above... I uploaded the same file (same name, size, contents) to two projects and it was counted against me twice for each file.

I don't have a way (that I know of) to manipulate the DB.

**Comment by GaryKriebel** [comment]
As discussed with @chrisseto ... he is cleaning up the proposal to ensure comments are reflected and resolved and then making sure the PR addresses the proposal.

**Comment by GaryKriebel** [comment]
While the proposal had the previous comments cleaned up, there is still confusion surrounding the requirements.... I have added to the proposal for further comment/clarification.

**Comment by sloria** [comment]
Closing for now, since we've decided to got with simple storage space auditing for the time being (implemented in https://github.com/CenterForOpenScience/osf.io/pull/4133)

**Pull request closed without merging**